modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/syncfs/SyncFsUtils.java

```diff
@@ -0,0 +1,146 @@
+     * According to {@code man}:
+     * <p/><i>syncfs() is like sync(), but synchronizes just the filesystem containing file referred to by the open file
+     * descriptor fd.<br/>
+     * ... sync() or syncfs() provide the same guarantees as fsync() called on every file in the system or
+     * filesystem respectively.</i>
+     *
+     * @param updStores Internal map from the checkpointer that contains all updated page stores.
+     * @throws IgniteCheckedException If operation failed.
+     */
+    public static void syncfs(
+        ConcurrentMap<PageStore, CheckpointPageStoreInfo> updStores
+    ) throws IgniteCheckedException {
+        IntHashMap<IgniteBiTuple<Path, FileStore>> groupIdToFileStore = new IntHashMap<>();
+
+        // First pass is to split all file stores by the groupId. All partitions within a group are always in the same
+        // directory.
+        for (Map.Entry<PageStore, CheckpointPageStoreInfo> entry : updStores.entrySet()) {
+            CheckpointPageStoreInfo info = entry.getValue();
+
+            if (groupIdToFileStore.containsKey(info.groupId))
+                continue;
+
+            Path path = ((FilePageStore) entry.getKey()).getPath();
+
+            try {
+                groupIdToFileStore.put(
+                    info.groupId,
+                    new IgniteBiTuple<>(path, Files.getFileStore(path))
+                );
+            }
+            catch (IOException e) {
+                throw new StorageException(e);
+            }
+        }
+
+        // Maps file store name to a path from that file store. Second pass is to fill this map.
+        Map<String, Path> fileSystems = new HashMap<>();
+
+        groupIdToFileStore.forEach((key, val) -> fileSystems.putIfAbsent(val.getValue().name(), val.getKey()));
+
+        // Resulting collection should not be big. For most installations it will only consist of a single element.
+        for (Path path : fileSystems.values()) {
+            // Open with default flags and mode in order to get file descriptor.
+            int fd = SyncFsLibC.open(path.toAbsolutePath().toString(), 0, 0);
+
+            int syncFsErr = 0;
+
+            try {
+                if (SyncFsLibC.syncfs(fd) != 0)
+                    syncFsErr = Native.getLastError();
+            } finally {
+                int closeErr = 0;
+
+                if (SyncFsLibC.close(fd) != 0)
+                    closeErr = Native.getLastError();
+
+                if (closeErr != 0 || syncFsErr != 0) {
+                    List<String> messages = new ArrayList<>(2);
+
+                    if (syncFsErr != 0)
+                        messages.add(SyncFsLibC.strerror(syncFsErr));
+
+                    if (closeErr != 0)
+                        messages.add(SyncFsLibC.strerror(closeErr));
+
+                    //noinspection ThrowFromFinallyBlock
+                    throw new IgniteCheckedException("Failed to execute syncfs: " + messages);
```

Comment:
Do we really need to throw from a `finally` block? Aren't we going to lose some throwable thrown from the `try` block?

Reply:
I'll think what I can do. If seems like I'll write my own try-with-resources